### PR TITLE
Master landed costs sle

### PR DESCRIPTION
**Pull request by sle-odoo**

### CODE REVIEW

File: addons/stock_landed_costs/models/account_move.py [hunk 1]
```diff
@@ -0,0 +1,73 @@
+# -*- coding: utf-8 -*-
+# Part of Odoo. See LICENSE file for full copyright and licensing details.
+
+from odoo import api, fields, models
+
+
+class AccountMove(models.Model):
+    _inherit = 'account.move'
+
+    landed_costs_ids = fields.One2many('stock.landed.cost', 'vendor_bill_id', string='Vendor Bills')
```
Comment (Whenrow) [inline addons/stock_landed_costs/models/account_move.py]: wrong string

File: addons/stock_landed_costs/models/account_move.py [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+            if self.product_type != 'service':
+                self.account_id = accounts['expense']
+                self.is_landed_costs_line = False
+            elif self.is_landed_costs_line and self.move_id.company_id.anglo_saxon_accounting:
+                self.account_id = accounts['stock_input']
+            else:
+                self.account_id = accounts['expense']
+
+    @api.onchange('product_id')
+    def _onchange_is_landed_costs_line_product(self):
```
Comment (Whenrow) [inline addons/stock_landed_costs/models/account_move.py]: why not a compute

Comment (Whenrow) [reply addons/stock_landed_costs/models/account_move.py]: please explain

File: addons/stock_landed_costs/views/stock_landed_cost_views.xml [hunk 3]
```diff
@@ -98,6 +97,8 @@
                             </page>
                         </notebook>
                     </sheet>
+
+
```
Comment (Whenrow) [inline addons/stock_landed_costs/views/stock_landed_cost_views.xml]: useless diff

File: addons/stock_account/__init__.py [hunk 4]
```diff
@@ -27,13 +27,13 @@ def _configure_journals(cr, registry):
...
                 ('type', '=', 'general')], limit=1).id
             if not journal_id:
-              journal_id = env['account.journal'].create({
-                'name': _('Inventory Valuation'),
-                'type': 'general',
-                'code': 'STJ',
-                'company_id': company_id.id,
-                'show_on_dashboard': False
-              }).id
+                journal_id = env['account.journal'].create({
```
Comment (Whenrow) [inline addons/stock_account/__init__.py]: useless diff

File: addons/stock_landed_costs/__init__.py [hunk 5]
```diff
@@ -1,4 +1,5 @@
 # -*- coding: utf-8 -*-
 # Part of Odoo. See LICENSE file for full copyright and licensing details.
 
-from . import models
\ No newline at end of file
+from . import models
```
Comment (Whenrow) [inline addons/stock_landed_costs/__init__.py]: useless diff

File: addons/stock_landed_costs/models/stock_landed_cost.py [hunk 6]
```diff
@@ -78,12 +78,15 @@ def button_cancel(self):
...
-        if any(not cost.valuation_adjustment_lines for cost in self):
-            raise UserError(_('No valuation adjustments lines. You should maybe recompute the landed costs.'))
-        if not self._check_sum():
-            raise UserError(_('Cost and adjustments lines do not match. You should maybe recompute the landed costs.'))
 
         for cost in self:
+            if not cost.picking_ids:
+                raise UserError(_('Please define the transfers on which those additional costs should apply.'))
+            if not cost.valuation_adjustment_lines:
+                cost.compute_landed_cost()
```
Comment (Whenrow) [inline addons/stock_landed_costs/models/stock_landed_cost.py]: Could you call compute_landed_cost on self directly ?

Comment (sle-odoo) [reply addons/stock_landed_costs/models/stock_landed_cost.py]: nope, as the user could manually change the generated lines before validating the lc

Comment (sle-odoo) [reply addons/stock_landed_costs/models/stock_landed_cost.py]: ok i could try to batch a little more

File: addons/stock_landed_costs/models/account_move.py (hunk 1 above)
Comment (sswapnesh) [inline addons/stock_landed_costs/models/account_move.py]: Maybe landed_costs_ids --> landed_cost_ids

File: addons/stock_landed_costs/__init__.py (hunk 5 above)
Comment (sswapnesh) [reply addons/stock_landed_costs/__init__.py]: > useless diff

Not really

File: addons/stock_landed_costs/models/account_move.py [hunk 7]
```diff
@@ -0,0 +1,73 @@
...
+    @api.depends('line_ids', 'line_ids.is_landed_costs_line')
+    def _compute_landed_costs_visible(self):
+        for account_move in self:
+            if account_move.landed_costs_ids:
+                account_move.landed_costs_visible = False
+            else:
+                account_move.landed_costs_visible = any(line.is_landed_costs_line for line in account_move.line_ids)
+
+    def button_create_landed_costs(self):
+        """Create a `stock.landed.costs` record associated to the account move of `self`, each
```
Comment (sswapnesh) [inline addons/stock_landed_costs/models/account_move.py]: stock.landed.cost

File: addons/stock_landed_costs/models/stock_landed_cost.py [hunk 8]
```diff
@@ -4,16 +4,38 @@
...
+]
+
+
 class LandedCost(models.Model):
     _name = 'stock.landed.cost'
     _description = 'Stock Landed Cost'
     _inherit = ['mail.thread', 'mail.activity.mixin']
 
+    def _default_account_journal_id(self):
+        """Take the journal of the company id defined, else fallback on the stock journal."""
```
Comment (sswapnesh) [inline addons/stock_landed_costs/models/stock_landed_cost.py]:
```suggestion
        """Take the journal configured in the company, else fallback on the stock journal."""
```

File: addons/stock_landed_costs/models/stock_landed_cost.py [hunk 9]
```diff
@@ -44,10 +66,13 @@ class LandedCost(models.Model):
         copy=False, readonly=True)
     account_journal_id = fields.Many2one(
         'account.journal', 'Account Journal',
-        required=True, states={'done': [('readonly', True)]})
+        required=True, states={'done': [('readonly', True)]}, default=_default_account_journal_id)
```
Comment (sswapnesh) [inline addons/stock_landed_costs/models/stock_landed_cost.py]:
```suggestion
        required=True, states={'done': [('readonly', True)]}, default=lambda self: self._default_account_journal_id())
```

File: addons/stock_landed_costs/models/stock_landed_cost.py [hunk 10]
```diff
@@ -44,10 +66,13 @@ class LandedCost(models.Model):
...
     company_id = fields.Many2one('res.company', string="Company",
         related='account_journal_id.company_id', readonly=False)
     stock_valuation_layer_ids = fields.One2many('stock.valuation.layer', 'stock_landed_cost_id')
+    vendor_bill_id = fields.Many2one(
+        'account.move', 'Vendor Bill', copy=False, domain=[('type', '=', 'in_invoice')])
+    currency_id = fields.Many2one('res.currency', related='company_id.currency_id', readonly=True)
```
Comment (sswapnesh) [inline addons/stock_landed_costs/models/stock_landed_cost.py]:
```suggestion
    currency_id = fields.Many2one('res.currency', related='company_id.currency_id')
```

File: addons/stock_landed_costs/models/stock_landed_cost.py [hunk 11]
```diff
@@ -282,28 +319,21 @@ class AdjustmentLines(models.Model):
...
-        digits=0, store=True)
+        'Original Value', digits='Product Price')
     additional_landed_cost = fields.Float(
         'Additional Landed Cost',
         digits='Product Price')
     final_cost = fields.Float(
-        'Final Cost', compute='_compute_final_cost',
+        'New Value', compute='_compute_final_cost',
         digits=0, store=True)
+    currency_id = fields.Many2one('res.currency', related='cost_id.company_id.currency_id', readonly=True)
```
Comment (sswapnesh) [inline addons/stock_landed_costs/models/stock_landed_cost.py]:
```suggestion
    currency_id = fields.Many2one('res.currency', related='cost_id.company_id.currency_id')
```

File: addons/stock_landed_costs/views/account_move_views.xml [hunk 12]
```diff
@@ -0,0 +1,33 @@
+<?xml version="1.0" encoding="utf-8"?>
+<odoo>
+    <record id="account_view_move_form_inherited" model="ir.ui.view">
```
Comment (sswapnesh) [inline addons/stock_landed_costs/views/account_move_views.xml]: Maybe 
```suggestion
    <record id="account_move_view_form_inherited" model="ir.ui.view">
```

File: addons/stock_landed_costs/views/res_config_settings_views.xml [hunk 13]
```diff
@@ -7,8 +7,11 @@
         <field name="inherit_id" ref="stock.res_config_settings_view_form"/>
         <field name="arch" type="xml">
             <div name="landed_cost_info" position="replace">
-                <div class="mt16" attrs="{'invisible': [('module_stock_landed_costs','=',False)]}">
-                    <button name="%(stock_landed_costs.stock_landed_cost_type_action)d" icon="fa-arrow-right" type="action" string="Landed Cost Types" class="btn-link"/>
+                <div class="content-group" attrs="{'invisible': [('module_stock_landed_costs','=',False)]}">
```
Comment (sswapnesh) [inline addons/stock_landed_costs/views/res_config_settings_views.xml]:
```suggestion
                <div class="content-group" attrs="{'invisible': [('module_stock_landed_costs', '=', False)]}">
```

File: addons/stock_landed_costs/models/account_move.py [hunk 14]
```diff
@@ -0,0 +1,73 @@
...
+
+
+class AccountMoveLine(models.Model):
+    _inherit = 'account.move.line'
+
+    product_type = fields.Selection(related='product_id.type', readonly=True)
+    is_landed_costs_line = fields.Boolean()
+
+    @api.onchange('is_landed_costs_line')
+    def _onchange_is_landed_costs_line(self):
```
Comment (Whenrow) [inline addons/stock_landed_costs/models/account_move.py]: I would keep only
```python
if self.is_landed_costs_line and self.move_id.company_id.anglo_saxon_accounting:
                self.account_id = accounts['stock_input']
```
and set a domain on is_landed_cost_line (produc_type, '=', 'service')

**Comment by nhomar** [comment]
@sle-odoo I know it is boring but can you set here the rationale?

cc @hbto aware!

**Comment by sle-odoo** [comment]
@nhomar Hi, if you're talking about the changes on the generated account move lines i'm writing a test that will explain the new flow. The idea is that at before this pr, the only way to balance the accounts was by setting the stock in account as expense account on the LC product. We also now give the possibility to link a LC with a vendor bill (and reconcile the entries).

fyi there's also [this change](https://github.com/odoo/odoo/commit/ee50b4a20c08cee4651597b7ad93d6e3bc26ccb7#diff-5337387cc67e910805d449c2d0186872) on the landed costs for v13.

**Comment by nhomar** [comment]
@sle-odoo I was talking more or less the rationale, but yes, the test can be explanatory enought a little description can complement it only.

thanks for the quick answer

File: addons/stock_account/__init__.py (hunk 4 above)
Comment (sle-odoo) [reply addons/stock_account/__init__.py]: it fixes the indent, i'd like to keep it

File: addons/stock_landed_costs/models/account_move.py (hunk 14 above)
Comment (sle-odoo) [reply addons/stock_landed_costs/models/account_move.py]: there's a domain in the view
> attrs="{'readonly': [('product_type', '!=', 'service')]}"

File: addons/stock_account/models/stock_picking.py [hunk 15]
```diff
@@ -11,5 +11,6 @@ def action_view_stock_valuation_layers(self):
         self.ensure_one()
         domain = [('id', 'in', self.move_lines.stock_valuation_layer_ids.ids)]
         action = self.env.ref('stock_account.stock_valuation_layer_action').read()[0]
-        return dict(action, domain=domain)
+        context = dict(self.env.context, no_at_date=True)
```
Comment (amoyaux) [inline addons/stock_account/models/stock_picking.py]: action['context']

File: addons/stock_landed_costs/views/account_move_views.xml [hunk 16]
```diff
@@ -0,0 +1,33 @@
...
+        <field name="name">account.view.move.form.inherited</field>
+        <field name="model">account.move</field>
+        <field name="inherit_id" ref="account.view_move_form"/>
```
Comment (amoyaux) [inline addons/stock_landed_costs/views/account_move_views.xml]: Check for customer invoice

**Comment by sle-odoo** [comment]
robodoo r+ rebase-ff

**Comment by sle-odoo** [comment]
robodoo r+ rebase-ff

**Comment by sle-odoo** [comment]
robodoo r+ rebase-ff

**Pull request closed without merging**